File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = [1] * n
        suffix = [1] * n
        output = [1] * n

        for i in range(n-1): 
            prefix[i+1] = prefix[i] * nums[i]

        for i in range(n-1,0,-1):
            suffix[i-1] = suffix[i] * nums[i]

        for i in range(n):
            output[i] = suffix[i] * prefix[i]
        return output

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int l = 0, r = n-1;
        int cut = 0;

        if (target == nums[r]) return r;

        while (l <= r) {
            int m = (r - l) / 2 + l;
            if (nums[m] > nums[r]) {
                l = m + 1;
            } else if (nums[m] < nums[r]) {
                r = m;
            } else {
                cut = m;
                break;
            }
        }
        r = n-1;
        if (target < nums[r]) {
            l = cut;
        } else  {
            l = 0;
            r = cut - 1;
        }

        while (l <= r) {
            int m = (r - l) / 2 + l;
            if (nums[m] < target) {
                l = m + 1;
            } else if (nums[m] > target) {
                r = m - 1;
            } else {
                return m;
            }
        }
        return -1;
    }
};
